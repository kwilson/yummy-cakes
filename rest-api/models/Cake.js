class Cake {
    constructor(id, name, imageUrl, yumFactor, comment) {
        this.id = id;
        this.name = name;
        this.comment = comment;
        this.imageUrl = imageUrl;
        this.yumFactor = yumFactor;
    }
}

module.exports = Cake;