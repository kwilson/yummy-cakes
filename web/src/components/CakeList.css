.cake-list {
    display: flex;
    flex-wrap: wrap;
    list-style-type: none;
    margin: 0 auto;
    padding: 0;
}

.cake-list > .cake-wrapper {
    margin: 5px;
}

