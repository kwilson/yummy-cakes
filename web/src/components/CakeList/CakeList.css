.cake-list {
    display: flex;
    flex-wrap: wrap;
    list-style-type: none;
    margin: 0 auto;
    padding: 0;
}

.cake-list > .cake-wrapper {
    background: url('/img/no-img.png') 50% 50% no-repeat;
    background-size: cover;
    margin: 5px;
}

