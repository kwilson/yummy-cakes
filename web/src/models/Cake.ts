export interface Cake {
  id: string;
  name: string;
  imageUrl: string;
  yumFactor: number;
  comment: string;
}
