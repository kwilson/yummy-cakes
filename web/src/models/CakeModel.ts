export interface CakeModel {
  id: string;
  name?: string;
  imageUrl?: string;
  yumFactor?: number;
  comment?: string;
}
