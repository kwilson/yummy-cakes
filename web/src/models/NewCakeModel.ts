export interface NewCakeModel {
  name?: string;
  imageUrl?: string;
  yumFactor?: number;
  comment?: string;
}
