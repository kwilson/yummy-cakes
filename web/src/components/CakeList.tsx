import * as React from 'react';
import { Cake } from '../models/Cake';

export interface CakeListStateProps {
  cakes: Cake[];
  isLoading: boolean;
}

export interface CakeListDispatchProps {
  selectCake: (id: string) => any;
}

const CakeList: React.StatelessComponent<CakeListStateProps & CakeListDispatchProps> = props => {
  const cakes = props.cakes.map(x => {
    const onClick = () => props.selectCake(x.id);
    return (
      <p onClick={onClick} key={x.id}>
        {x.name}
      </p>
    );
  });

  return <div>{cakes}</div>;
};

export default CakeList;
