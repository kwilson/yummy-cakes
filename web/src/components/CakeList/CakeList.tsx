import * as React from 'react';

import { Cake } from '../';
import { CakeModel } from '../../models/CakeModel';

import './CakeList.css';

export interface CakeListProps {
  cakes: CakeModel[];
  onSelectCake: (id: string) => any;
}

const CakeList: React.StatelessComponent<CakeListProps> = props => {
  const getCakes = (cakes: CakeModel[]) => cakes.map(x => {
    const onClick = () => props.onSelectCake(x.id);

    return (
      <li key={x.id} className="cake-wrapper">
        <Cake
          cake={x}
          selectCake={onClick}
        />
      </li>
    );
  });

  if (!props.cakes || props.cakes.length < 1) {
    return (
      <div className="cake-list--empty">No cakes to view</div>
    );
  }

  return (
    <ul className="cake-list">
      {getCakes(props.cakes)}
    </ul>
  );
};

export default CakeList;
