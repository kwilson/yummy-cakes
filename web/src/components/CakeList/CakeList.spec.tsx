import * as React from 'react';
import { shallow } from 'enzyme';
import { create as createSnapshot } from 'react-test-renderer';

import CakeList from './CakeList';
import { CakeModel } from '../../models/CakeModel';

import { Cake } from '../';

describe('CakeList Component', () => {
  let onSelectCakeHandler: jest.Mock<{}>;
  beforeEach(() => {
    onSelectCakeHandler = jest.fn();
  });

  it('displays placeholder text if there are no cakes', () => {
    // Arrange
    const cakes: CakeModel[] = [];

    const cmp = <CakeList cakes={cakes} onSelectCake={onSelectCakeHandler} />;

    // Act
    const el = shallow(cmp);
    const snapshot = createSnapshot(cmp).toJSON();

    // Assert
    expect(el.find('.cake-list')).toHaveLength(0);

    expect(el.find('.cake-list--empty').text()).toEqual('No cakes to view');
    expect(snapshot).toMatchSnapshot();
  });

  it('retuns a child component for each cake', () => {
    // Arrange
    const cakes: CakeModel[] = [1, 2, 3, 4, 5].map(value => ({
      id: `cake-${value}`,
      name: `Cake Name ${value}`
    }));

    const cmp = <CakeList cakes={cakes} onSelectCake={onSelectCakeHandler} />;

    // Act
    const el = shallow(cmp);
    const snapshot = createSnapshot(cmp).toJSON();

    // Assert
    const cakeComponents = el.find(Cake);
    expect(cakeComponents).toHaveLength(cakes.length);

    cakeComponents.forEach((cakeComponent, idx) => {
      expect(cakeComponent.props().cake).toBe(cakes[idx]);

      cakeComponent.props().selectCake();
      expect(onSelectCakeHandler).toHaveBeenCalledTimes(1);
      expect(onSelectCakeHandler).toHaveBeenCalledWith(cakes[idx].id);

      jest.resetAllMocks();
    });

    expect(snapshot).toMatchSnapshot();
  });
});
