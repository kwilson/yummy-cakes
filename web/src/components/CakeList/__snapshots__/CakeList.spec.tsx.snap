// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CakeList Component displays placeholder text if there are no cakes 1`] = `
<div
  className="cake-list--empty"
>
  No cakes to view
</div>
`;

exports[`CakeList Component retuns a child component for each cake 1`] = `
<ul
  className="cake-list"
>
  <li
    className="cake-wrapper"
  >
    <button
      className="cake"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    >
      <div
        className="cake__details"
      >
        Cake Name 1
      </div>
    </button>
  </li>
  <li
    className="cake-wrapper"
  >
    <button
      className="cake"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    >
      <div
        className="cake__details"
      >
        Cake Name 2
      </div>
    </button>
  </li>
  <li
    className="cake-wrapper"
  >
    <button
      className="cake"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    >
      <div
        className="cake__details"
      >
        Cake Name 3
      </div>
    </button>
  </li>
  <li
    className="cake-wrapper"
  >
    <button
      className="cake"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    >
      <div
        className="cake__details"
      >
        Cake Name 4
      </div>
    </button>
  </li>
  <li
    className="cake-wrapper"
  >
    <button
      className="cake"
      onClick={[Function]}
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    >
      <div
        className="cake__details"
      >
        Cake Name 5
      </div>
    </button>
  </li>
</ul>
`;
